### PR TITLE
test: migrate binary_name to snapbox

### PR DESCRIPTION
**Pull request by weihanglo**
### What does this PR try to resolve?

Part of <https://github.com/rust-lang/cargo/issues/14039>.

Migrate `tests/testsuite/binary_name.rs` to snapbox.

### How should we test and review this PR?

Two things need to beware of

* A new redaction is added: `[ELAPSED]` for libtest output.
* Snapbox's [`JsonLines`](https://docs.rs/snapbox/0.6.10/snapbox/data/enum.DataFormat.html#variant.JsonLines) doesn't seems to support pretty printed JSON.

### CODE REVIEW

File: tests/testsuite/binary_name.rs [hunk 1]
```diff
@@ -284,25 +312,13 @@ fn check_msg_format_json() {
...
-
-{"reason":"build-finished", "success":true}
-"#;
-
     // Run cargo build.
     p.cargo("build --message-format=json")
         .masquerade_as_nightly_cargo(&["different-binary-name"])
-        .with_json(output)
+        .with_stdout_data(str![[r#"
+{"executable":"[ROOT]/foo/target/debug/007bar[EXE]","features":[],"filenames":"{...}","fresh":false,"manifest_path":"[ROOT]/foo/Cargo.toml","package_id":"path+[ROOTURL]/foo#0.0.1","profile":"{...}","reason":"compiler-artifact","target":"{...}"}
```
Comment (weihanglo) [inline tests/testsuite/binary_name.rs:319]: This is something worth fixing in snapbox. We need a pretty-printed JSON lines support.

Comment (epage) [reply tests/testsuite/binary_name.rs:319]: Technically, this is no longer jsonlines.  It looks like cargo-test-support had a baked in policy of taking `expected` and using double-newlines to delimit pretty-printed json.

One trick I can look into is we could parse a json array and then convert that to jsonlines.

**Comment by epage** [comment]
@bors r+

Checked in with @weihanglo  on jsonlines

> It's a regression when reviewing those outputs. We can track it in the issue and revisit then

File: tests/testsuite/binary_name.rs (hunk 1 above)
Comment (weihanglo) [reply tests/testsuite/binary_name.rs:319]: Yeah I am aware of the double-newlines trick in cargo-test-support. I don't consider it as a block for this test file. For test suites like `tests/testsuite/message_format.rs` we should consider postponing the migrations.

**Pull request merged**